File: day1.py
from collections import Counter

left: list[int] = []
right: list[int] = []

with open("2024-1.txt", "r") as f:
    for line in f:
        left.append(int(line.split()[0]))
        right.append(int(line.split()[1]))

left.sort()
right.sort()

sum3 = 0
for l, r in zip(left, right):
    sum3 += l - r if l - r >= 0 else r - l

print(sum3)

leftset = set(left)

rightmap = Counter(right)

sum2 = 0

for x in leftset:
    if x in rightmap:
        sum2 += x * rightmap[x]

print(sum2)

similarity = 0

for element in left:
    similarity += element * sum(1 for n in right if n == element)

print(f"{similarity=}")

File: src/bin/day1.rs
use std::{
    collections::{HashMap, HashSet},
    fs::File,
    io::read_to_string,
};

fn calculate(filename: &str) {
    let mut left: Vec<u32> = Vec::new();
    let mut right: Vec<u32> = Vec::new();

    let file = File::open(filename).unwrap();

    for line in read_to_string(file).unwrap().lines() {
        left.push(line.split_once("   ").unwrap().0.parse().unwrap());
        right.push(line.split_once("   ").unwrap().1.parse().unwrap());
    }

    left.sort();
    right.sort();

    let mut distance = 0;
    for (index, element) in left.iter().enumerate() {
        distance += element.abs_diff(right[index]);
    }
    println!("{}", distance);

    let mut distance2 = 0;

    let mut left2 = HashSet::new();
    for num in &left {
        left2.insert(num);
    }
    let mut right2 = HashMap::new();
    for num in &right {
        right2.entry(num).and_modify(|val| *val += 1).or_insert(1);
    }

    for num in left2 {
        if right2.contains_key(&num) {
            distance2 += num * right2[&num];
        }
    }

    println!("{}", distance2);

    let mut similarity = 0;

    for element in left {
        similarity += element * (right.iter().filter(|&n| *n == element).count() as u32);
    }
    println!("{}", similarity);
}

fn main() {
    calculate("./2024-1.txt");
}
